google/cloud/aiplatform/prediction/local_model.py

```diff
@@ -0,0 +1,634 @@
+            |-- user_code/
+            |   |-- utils.py
+            |   |-- custom_package.tar.gz
+            |   |-- ...
+            |-- ...
+            |-- entrypoint.py
+
+        Args:
+            src_dir (str):
+                Required. The path to the local directory including all needed files such as
+                predictor. The whole directory will be copied to the image.
+            output_image_uri (str):
+                Required. The image uri of the built image.
+            predictor (Type[Predictor]):
+                Optional. The custom predictor class consumed by handler to do prediction.
+            handler (Type[Handler]):
+                Required. The handler class to handle requests in the model server.
+            base_image (str):
+                Required. The base image used to build the custom images. The base image must
+                have python and pip installed where the two commands `python` and `pip` must be
+                available.
+            requirements_path (str):
+                Optional. The path to the local requirements.txt file. This file will be copied
+                to the image and the needed packages listed in it will be installed.
+            extra_packages (List[str]):
+                Optional. The list of user custom dependency packages to install.
+            no_cache (bool):
+                Required. Do not use cache when building the image. Using build cache usually
+                reduces the image building time. See
+                https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#leverage-build-cache
+                for more details.
+
+        Returns:
+            local model: Instantiated representation of the local model.
+        """
+        entrypoint_file = "entrypoint.py"
+
+        prediction_utils.populate_entrypoint_if_not_exists(
+            src_dir,
+            entrypoint_file,
+            predictor=predictor,
+            handler=handler,
+        )
+
+        is_prebuilt_prediction_image = helpers.is_prebuilt_prediction_container_uri(
+            base_image
+        )
+        _ = build.build_image(
+            base_image,
+            src_dir,
+            Path(src_dir).joinpath(entrypoint_file).as_posix(),
+            output_image_uri,
+            requirements_path=requirements_path,
+            extra_requirements=_DEFAULT_SDK_REQUIREMENTS,
+            extra_packages=extra_packages,
+            exposed_ports=[DEFAULT_HTTP_PORT],
+            pip_command="pip3" if is_prebuilt_prediction_image else "pip",
+            python_command="python3" if is_prebuilt_prediction_image else "python",
+            no_cache=no_cache,
+        )
+
+        container_spec = gca_model_compat.ModelContainerSpec(
+            image_uri=output_image_uri,
+            predict_route=DEFAULT_PREDICT_ROUTE,
+            health_route=DEFAULT_HEALTH_ROUTE,
+        )
+
+        return cls(container_spec)
+
+    def upload(
```

Comment:
I would strongly recommend making the LocalModel class a subclass of Model and using Model's `upload` method. The method signature for `upload` will change (significantly, in the near future), and hardcoding upload like this just makes maintaining compatibility with Model.upload more difficult.

Reply:
The prediction README says that LocalModel is designed to be analogous to Model. In that case, please base this class on Model and override behavior as necessary.

Reply:
We considered this but we decided not to make LocalModel class a subclass of Model since LocalModel hasn't had a Vertex Model resource created yet. It's like "the local version" of LocalModel and the only way to bridge LocalModel to Model is the `upload` function (see https://github.com/googleapis/python-aiplatform/blob/custom-prediction-routine/google/cloud/aiplatform/prediction/docs/cpr_cuj.svg)

Reply:
I'm not sure I'm understanding this explanation. If LocalModel is the local version of Model, you should make it a subclass of Model, use the base class's `upload` implementation, and override other relevant methods.

This is essential, we can't support multiple different model interfaces.

Reply:
A compromise that should be easier for you, without overhauling too much: Please create mixins, something like ModelMixin and EndpointMixin. Model and LocalModel should both inherit ModelMixin, Endpoint and LocalEndpoint should both inherit EndpointMixin.

The mixins should cover all public methods for Model and Endpoint, without any behavior changes for Model and Endpoint.